libraries/chain/db_management.cpp

```diff
@@ -46,27 +46,36 @@ database::~database()
    clear_pending();
 }
 
-void database::reindex(fc::path data_dir, const genesis_state_type& initial_allocation)
+void database::reindex( fc::path data_dir )
 { try {
-   ilog( "reindexing blockchain" );
-   wipe(data_dir, false);
-   open(data_dir, [&initial_allocation]{return initial_allocation;});
-
-   auto start = fc::time_point::now();
    auto last_block = _block_id_to_block.last();
    if( !last_block ) {
       elog( "!no last block" );
       edump((last_block));
       return;
    }
+   if( last_block->block_num() <= head_block_num()) return;
 
+   ilog( "reindexing blockchain" );
+   auto start = fc::time_point::now();
    const auto last_block_num = last_block->block_num();
+   uint32_t flush_point = last_block_num - 10000;
+   uint32_t undo_point = last_block_num - 50;
 
    ilog( "Replaying blocks..." );
-   _undo_db.disable();
-   for( uint32_t i = 1; i <= last_block_num; ++i )
+   if( head_block_num() >= undo_point )
+      _fork_db.start_block( *fetch_block_by_number( head_block_num() ) );
```

Comment:
The original code doesn't call _fork_db.start_block in all cases either. AFAICS fork_db uses the first pushed block as the start block if no start block has been set explicitly.